// electron-builder/out/targets/nsis.js
"use strict";

var _slicedToArray = function () { function sliceIterator(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"]) _i["return"](); } finally { if (_d) throw _e; } } return _arr; } return function (arr, i) { if (Array.isArray(arr)) { return arr; } else if (Symbol.iterator in Object(arr)) { return sliceIterator(arr, i); } else { throw new TypeError("Invalid attempt to destructure non-iterable instance"); } }; }();

const metadata_1 = require("../metadata");
const util_1 = require("../util/util");
const path = require("path");
const bluebird_1 = require("bluebird");
const binDownload_1 = require("../util/binDownload");
const uuid_1345_1 = require("uuid-1345");
const platformPackager_1 = require("../platformPackager");
const archive_1 = require("./archive");
const log_1 = require("../util/log");
const fs_extra_p_1 = require("fs-extra-p");
const semver = require("semver");
//noinspection JSUnusedLocalSymbols
const __awaiter = require("../util/awaiter");
const NSIS_VERSION = "nsis-3.0.0";
//noinspection SpellCheckingInspection
const NSIS_SHA2 = "7741089f3ca13de879f87836156ef785eab49844cacbeeabaeaefd1ade325ee7";
//noinspection SpellCheckingInspection
const ELECTRON_BUILDER_NS_UUID = "50e065bc-3134-11e6-9bab-38c9862bdaf3";
const nsisPathPromise = binDownload_1.getBin("nsis", NSIS_VERSION, `https://dl.bintray.com/electron-userland/bin/${ NSIS_VERSION }.7z`, NSIS_SHA2);
class NsisTarget extends platformPackager_1.Target {
    constructor(packager, outDir) {
        super("nsis");
        this.packager = packager;
        this.outDir = outDir;
        this.archs = new Map();
        this.options = packager.info.devMetadata.build.nsis || Object.create(null);
    }
    build(arch, appOutDir) {
        return __awaiter(this, void 0, void 0, function* () {
            const packager = this.packager;
            const archSuffix = metadata_1.Arch[arch];
            const archiveFile = path.join(this.outDir, `${ packager.appInfo.name }-${ packager.appInfo.version }-${ archSuffix }.nsis.7z`);
            this.archs.set(arch, log_1.task(`Creating NSIS ${ archSuffix } package`, archive_1.archiveApp(packager.devMetadata.build.compression, "7z", archiveFile, appOutDir, true)));
        });
    }
    finishBuild() {
        return log_1.task("Building NSIS installer", this.buildInstaller().then(() => bluebird_1.Promise.map(this.archs.values(), it => fs_extra_p_1.unlink(it))));
    }
    buildInstaller() {
        return __awaiter(this, void 0, void 0, function* () {
            const packager = this.packager;
            const iconPath = yield packager.getIconPath();
            const appInfo = packager.appInfo;
            const version = appInfo.version;
            const installerPath = path.join(this.outDir, `${ appInfo.productFilename } Setup ${ version }.exe`);
            const guid = this.options.guid || (yield bluebird_1.Promise.promisify(uuid_1345_1.v5)({ namespace: ELECTRON_BUILDER_NS_UUID, name: appInfo.id }));
            const defines = {
                APP_ID: appInfo.id,
                APP_GUID: guid,
                PRODUCT_NAME: appInfo.productName,
                PRODUCT_FILENAME: appInfo.productFilename,
                APP_DESCRIPTION: appInfo.description,
                VERSION: version,
                MUI_ICON: iconPath,
                MUI_UNICON: iconPath,
                COMPANY_NAME: appInfo.companyName,
                PROJECT_DIR: this.packager.projectDir,
                BUILD_RESOURCES_DIR: this.packager.buildResourcesDir
            };
            for (let _ref of this.archs) {
                var _ref2 = _slicedToArray(_ref, 2);

                let arch = _ref2[0];
                let file = _ref2[1];

                defines[arch === metadata_1.Arch.x64 ? "APP_64" : "APP_32"] = yield file;
            }
            const oneClick = this.options.oneClick !== false;
            const installerHeader = oneClick ? null : yield this.getResource(this.options.installerHeader, "installerHeader.bmp");
            if (installerHeader != null) {
                defines.MUI_HEADERIMAGE = null;
                defines.MUI_HEADERIMAGE_RIGHT = null;
                defines.MUI_HEADERIMAGE_BITMAP = installerHeader;
            }
            const installerHeaderIcon = oneClick ? yield this.getResource(this.options.installerHeaderIcon, "installerHeaderIcon.ico") : null;
            if (installerHeaderIcon != null) {
                defines.HEADER_ICO = installerHeaderIcon;
            }
            if (this.options.perMachine === true) {
                defines.INSTALL_MODE_PER_ALL_USERS = null;
            }
            if (!oneClick || this.options.perMachine === true) {
                defines.INSTALL_MODE_PER_ALL_USERS_REQUIRED = null;
            }
            if (oneClick) {
                if (this.options.runAfterFinish !== false) {
                    defines.RUN_AFTER_FINISH = null;
                }
            } else if (this.options.allowElevation !== false) {
                defines.MULTIUSER_INSTALLMODE_ALLOW_ELEVATION = null;
            }
            // Error: invalid VIProductVersion format, should be X.X.X.X
            // so, we must strip beta
            const parsedVersion = new semver.SemVer(appInfo.version);
            const commands = {
                OutFile: `"${ installerPath }"`,
                // LoadLanguageFile: '"${NSISDIR}/Contrib/Language files/English.nlf"',
                VIProductVersion: `${ parsedVersion.major }.${ parsedVersion.minor }.${ parsedVersion.patch }.${ appInfo.buildNumber || "0" }`,
                // VIFileVersion: packager.appInfo.buildVersion,
                VIAddVersionKey: [`ProductName "${ appInfo.productName }"`, `CompanyName "${ appInfo.versionString.CompanyName }"`, `LegalCopyright "${ appInfo.versionString.LegalCopyright }"`, `FileDescription "${ appInfo.description }"`, `FileVersion "${ appInfo.buildVersion }"`]
            };
            if (packager.devMetadata.build.compression === "store") {
                commands.SetCompress = "off";
                defines.COMPRESS = "off";
            } else {
                commands.SetCompressor = "lzma";
                // default is 8: test app installer size 37.2 vs 36 if dict size 64
                commands.SetCompressorDictSize = "64";
                defines.COMPRESS = "auto";
            }
            if (oneClick) {
                defines.ONE_CLICK = null;
            }
            util_1.debug(defines);
            util_1.debug(commands);
            if (this.packager.options.effectiveOptionComputed != null && this.packager.options.effectiveOptionComputed([defines, commands])) {
                return;
            }
            yield log_1.subTask(`Executing makensis`, this.executeMakensis(defines, commands));
            yield packager.sign(installerPath);
            this.packager.dispatchArtifactCreated(installerPath, `${ appInfo.name }-Setup-${ version }.exe`);
        });
    }
    getResource(custom, name) {
        return __awaiter(this, void 0, void 0, function* () {
            let result = custom;
            if (result === undefined) {
                const resourceList = yield this.packager.resourceList;
                if (resourceList.indexOf(name) !== -1) {
                    return path.join(this.packager.buildResourcesDir, name);
                }
            } else {
                return path.resolve(this.packager.projectDir, result);
            }
            return null;
        });
    }
    executeMakensis(defines, commands) {
        return __awaiter(this, void 0, void 0, function* () {
            const args = ["-WX"];
            for (let name of Object.keys(defines)) {
                const value = defines[name];
                if (value == null) {
                    args.push(`-D${ name }`);
                } else {
                    args.push(`-D${ name }=${ value }`);
                }
            }
            for (let name of Object.keys(commands)) {
                const value = commands[name];
                if (Array.isArray(value)) {
                    for (let c of value) {
                        args.push(`-X${ name } ${ c }`);
                    }
                } else {
                    args.push(`-X${ name } ${ value }`);
                }
            }
            const nsisTemplatesDir = path.join(__dirname, "..", "..", "templates", "nsis");
            args.push("-");
            const binDir = process.platform === "darwin" ? "mac" : process.platform === "win32" ? "Bin" : "linux";
            const nsisPath = yield nsisPathPromise;
            const packager = this.packager;
            const fileAssociations = asArray(packager.devMetadata.build.fileAssociations).concat(asArray(packager.platformSpecificBuildOptions.fileAssociations));
            let registerFileAssociationsScript = "";
            let unregisterFileAssociationsScript = "";
            let script = yield fs_extra_p_1.readFile((yield this.getResource(this.options.script, "installer.nsi")) || path.join(nsisTemplatesDir, "installer.nsi"), "utf8");
            const customInclude = yield this.getResource(this.options.include, "installer.nsh");
            if (customInclude != null) {
                script = `!include "${ customInclude }"\n!addincludedir "${ this.packager.buildResourcesDir }"\n${ script }`;
            }
            if (fileAssociations.length !== 0) {
                script = "!include FileAssociation.nsh\n" + script;
                for (let item of fileAssociations) {
                    registerFileAssociationsScript += '${RegisterExtension} "$INSTDIR\\${APP_EXECUTABLE_FILENAME}" ' + `"${ normalizeExt(item.ext) }" "${ item.name }"\n`;
                }
                for (let item of fileAssociations) {
                    unregisterFileAssociationsScript += "${UnRegisterExtension} " + `"${ normalizeExt(item.ext) }" "${ item.name }"\n`;
                }
            }
            script = script.replace("!insertmacro registerFileAssociations", registerFileAssociationsScript);
            script = script.replace("!insertmacro unregisterFileAssociations", unregisterFileAssociationsScript);
            if (util_1.debug.enabled) {
                process.stdout.write("\n\nNSIS script:\n\n" + script + "\n\n---\nEnd of NSIS script.\n\n");
            }
            yield new bluebird_1.Promise((resolve, reject) => {
                const command = path.join(nsisPath, binDir, process.platform === "win32" ? "makensis.exe" : "makensis");
                const childProcess = util_1.doSpawn(command, args, {
                    // we use NSIS_CONFIG_CONST_DATA_PATH=no to build makensis on Linux, but in any case it doesn't use stubs as MacOS/Windows version, so, we explicitly set NSISDIR
                    env: Object.assign({}, process.env, { NSISDIR: nsisPath }),
                    cwd: nsisTemplatesDir
                }, true);
                util_1.handleProcess("close", childProcess, command, resolve, reject);
                childProcess.stdin.end(script);
            });
        });
    }
}
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = NsisTarget;
// nsis — add leading dot, mac — remove leading dot
function normalizeExt(ext) {
    return ext.startsWith(".") ? ext : `.${ ext }`;
}
function asArray(v) {
    if (v == null) {
        return [];
    } else if (Array.isArray(v)) {
        return v;
    } else {
        return [v];
    }
}
